Introducing Extension Manager and Extensions

sophron opened the pull request:
This is Work In Progress. Will be updated.

anakin1028 commented on wifiphisher/pywifiphisher.py:
how about using `namedtuple` instead of `dictionary`? like the following:
`shared_data = namedtuple('ApData', 'target_channel', 'target_essid',)`

and then used as follows:

```python
shared_data.target_channel
shared_data.target_essid
```
will it be cleaner than dictionary?

anakin1028 commented on wifiphisher/extensions/deauth.py:
```diff
@@ -0,0 +1,137 @@
+                    # create and add deauthentication packets for client
+                    deauth_pkts += self._craft_packet(receiver, self._data["target_ap_bssid"])
+                    deauth_pkts += self._craft_packet(self._data["target_ap_bssid"], receiver)
+
```
Still need to consider the `--essid` case since in this case deauth is sending for all the clients and aps.
I already have a PR for that, just for reference here #600 :)

sophron replied:
Nice! The "frenzy" mode should also work with channel hopping though. I will implement the channel hopping for the whole Extension Manager after we merge this PR. After the Extension Manager supports channel hopping, you should be able to leverage that for the "frenzy" mode.

anakin1028 replied:
@sophron yes! Since for frenzy mode, we just need to change the address 3 to the bssid of the receiving frame, so it definitely can work with channel hopping feature. I can also wait for this PR done, and leverage on the `share_data.args` to see if the `--essid` is used and do a small modification about my PR.

sophron commented on wifiphisher/pywifiphisher.py:
Yes, `namedtuple` may turn out better.

The plan is to have a `shared_data` that will be shared among all extensions *and* templates. Currently, we feed templates with kinda different data but should be the same as in the extensions.

sophron commented on wifiphisher/pywifiphisher.py:
The only thing with `namedtuple` is that it is immutable. We may need to update our `shared_data` though to send more (fresher) information to our extensions.

For that reason, in my new commits I stick to `dictionary` but I do convert this to `namedtuple` before passing to extensions so that they can use the clearer syntax.

sophron commented:
I pushed new commits, including some tests. If there are no objections, I will merge this tomorrow.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+import os
```
Un-used import.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+import os
+import importlib
+import constants
```
Relative import. Should be `wifiphisher.common.constants` .

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+import scapy.layers.dot11 as dot11
+import scapy.arch.linux as linux
+
+class ExtensionManager():
```
Needs another empty line before it.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+
+class ExtensionManager():
+    """
+    Extension Manager (EM) defines an API for modular 
```
Multiple trailing whitespaces throughout this docstring.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+
+    def __init__(self):
+        """
+        Init the EM object. 
```
Trailing whitespace.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager
+        :param interface: String interface object
```
Doesn't have interface or shard_data variables.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        self._interface = None
+        self._socket = None
+        self._should_continue = True
+        self._packets_to_send = [] 
```
Trailing whitespace.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        self._should_continue = True
+        self._packets_to_send = [] 
+
+    def set_interface(self, interface):
```
This would be a lot better if it was a property rather than a method.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+
+    def init_extensions(self, shared_data):
+        """
+        Init EM extensions. Should be run 
```
Multiple trailing whitespaces.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        shared_data = collections.namedtuple('GenericDict', \
+                      shared_data.keys())(**shared_data)
+        # Initialize all extensions with the shared data
+        for m in constants.EXTENSIONS:
```
`m` is too short an unclear

anakin1028 replied:
Here the definition of `constants.EXTENSIONS = ["deauth", "lure10"]`, this means that when user uses the deauth interface, we will start lure10 attack. Is this what you want? Since I remember will have `--lure10` option for users before to indicate whether they want to do this attack or not.

anakin1028 replied:
I think we should have a `is_mod_used` variable to indicate whether the attack is being applied?

sophron replied:
@anakin1028 The extension will load, yes. But it won't do anything unless the user has specified it in the arguments (see `lure10.py`).

Maybe it's better if we don't load the extension if we aren't going to use it.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        # Initialize all extensions with the shared data
+        for m in constants.EXTENSIONS:
+            mod = importlib.import_module(constants.EXTENSIONS_LOADPATH + m)
+            Class = getattr(mod, m.title())
```
`Class` is an invalid variable name.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+                      shared_data.keys())(**shared_data)
+        # Initialize all extensions with the shared data
+        for m in constants.EXTENSIONS:
+            mod = importlib.import_module(constants.EXTENSIONS_LOADPATH + m)
```
`mod` is not as clear as it could be.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        for m in constants.EXTENSIONS:
+            mod = importlib.import_module(constants.EXTENSIONS_LOADPATH + m)
+            Class = getattr(mod, m.title())
+            obj = Class(shared_data)
```
`obj` is not as clear as it could be.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        """
+
+        # Convert shared_data from dict to named tuple
+        shared_data = collections.namedtuple('GenericDict', \
```
why are we converting a dictionary to a tuple here?

sophron replied:
We want the extensions to be able to use the notation `data.key` instead of `data["key"]`. See my discussion with @anakin1028 in a previous comment.

blackHatMonkey replied:
I'm all for using named tuples but in the previous comment you mentioned that the data might change and thats why you opted for a dictionary but what data needs to change?

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+
+    def start_extensions(self):
+        """
+        Starts the two main daemons of EM:
```
Trailing whitespace.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        self._interface = interface
+        self._socket = linux.L2Socket(iface=self._interface)
+
+    def init_extensions(self, shared_data):
```
Maybe we can come up with a better name?

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        """
+
+        output = []
+        for m in self._extensions:
```
perhaps we can use better variable names here as well.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+
+        # continue to find clients until told otherwise
+        while self._should_continue:
+            dot11.sniff(iface=self._interface, prn=self._process_packet,
```
Can't we just assign it to a variable and do deliver them to each module here instead of in `_process_packet`?

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        to _process_packet.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager 
```
Trailing whitespace.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        crafted by extensions.
+
+        :param self: An ExtensionManager object
+        :type self: ExtensionManager 
```
Trailing white space.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        :rtype: None
+        """
+
+        while self._should_continue:
```
Multiple loops with the same condition?

sophron replied:
Yes. We want to keep checking if there are any packets to send (first loop). 

If there are, we want to keep broadcasting packets (second loop). However in the second loop we don't want to include the first "if statement" for performance reasons.

blackHatMonkey replied:
why are we even checking the length of it. The for loop will take care of this. As an example:
```python
a = []
for item in a:
    print "HI"
```
will not run since there are no items in there.

blackHatMonkey commented on wifiphisher/extensions/deauth.py:
```diff
@@ -0,0 +1,137 @@
+
+        disassoc_packet = (dot11.RadioTap() / dot11.Dot11(type=0, subtype=10, \
+                    addr1=receiver, addr2=sender, addr3=self._data.target_ap_bssid) \
+                  / dot11.Dot11Disas())  
```
Trailing whitespace.

blackHatMonkey commented on wifiphisher/extensions/deauth.py:
```diff
@@ -0,0 +1,137 @@
+        :rtype: None
+        """
+
+        self._should_continue = False
```
This is a pointless statement here. In fact on_exit is not needed in here.

sophron replied:
You are correct that `_should_continue` should not exist here.

I'm still thinking if the idea of having separate `on_exit` callbacks for each extension makes sense. While in these two modules (deauth and lure10) there's no reason to do something, there may be cases where the scripts will want to do some cleanup on their own (or not).

blackHatMonkey replied:
When that issue arises we can implement it then but it should be removed for now.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Un-used import.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Un-used import.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Un-used import.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Un-used import.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Needs a extra line before it.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Trailing whitespace.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Lines are too long.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
would be much simpler with `string.format`.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
unreachable code.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Nothing here?

anakin1028 commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,194 @@
+        :return: None
+        :rtype: None
+        """
+
```
should join the send_thread and listen_thread as before.

sophron commented:
Thank you both for the review. If you can also do a functional testing and verify that everything works as expected, that would be great.

sophron commented:
Guys,

I addressed most of your comments. I also implemented channel hopping capabilities. There's still some work to be done but we are getting there.

blackHatMonkey commented:
@sophron Can you please fix the conflict?

sophron commented:
@blackHatMonkey @anakin1028 Thank you both the review. If you have any more comments, lemme know. Otherwise, I'll merge this tomorrow.

anakin1028 commented on wifiphisher/common/extensions.py:
```diff
@@ -211,14 +260,14 @@ def _send(self):
-                            self._socket.send(pkt)
+                for pkt in self._packets_to_send[self._current_channel] + \
+                        self._packets_to_send["*"]:
+                    self._socket.send(pkt)
```
what's the intention for `self._packets_to_send`?

anakin1028 commented:
@sophron shouldn't we wait for #581 done? Since lots of code will base on the `NetworkManager` and `NetworkAdaptor` classes. I just worry about #581 cannot be converged.

sophron commented:
Yes, we can do that if @blackHatMonkey agrees.

anakin1028 commented on wifiphisher/pywifiphisher.py:
```diff
@@ -549,6 +549,7 @@ def __init__(self):
         self.template_manager = phishingpage.TemplateManager()
         self.access_point = accesspoint.AccessPoint()
         self.fw = firewall.Fw()
+        self.em = extensions.ExtensionManager()
```
@sophron you should do `em` exit before calling `network_manager.on_exit`. `network_manager.on_exit` will set the card back to managed mode, thus will down the card. This will cause exception for scapy.

anakin1028 commented on wifiphisher/pywifiphisher.py:
```diff
@@ -772,12 +774,24 @@ def start(self):
             # set the channel on the deauthenticating interface
```
`monchannel` since redundant?

anakin1028 commented on wifiphisher/extensions/deauth.py:
```diff
@@ -0,0 +1,151 @@
+        :return: A list with all interested channels.
+        :rtype: list
+        """
+
```
Though you have implement the channel hopping, now the code seems only stay in the same target AP channel. Maybe after this PR is merged, I'll do channel hopping leveraging on your function for the `--essid` case.

sophron replied:
I implemented channel hopping in the `ExtensionManager`, the `deauth` module needs to leverage that now. Feel free to make a patch after this gets merged. ;)

blackHatMonkey commented:
@sophron Do you want to start review for this PR first or #600?

sophron commented:
As soon as #610 gets merged, this PR should follow. @anakin1028 can apply his patch (#600) on the new deauth code (which is now an extension).

blackHatMonkey commented:
@sophron Is this PR ready to be reviewed?

blackHatMonkey commented:
@sophron Also can you push all your changes in a branch inside `wifiphisher` as it easier to modify if needed.

sophron commented:
@blackHatMonkey There are some conflicts that I need to resolve but feel free to start reviewing.

blackHatMonkey commented:
This is list of all the stuff that needs to be reviewed. @anakin1028 feel free to update it as we go.

- [ ] test_exception
- [ ]
- [ ]
- [ ]
- [ ]
- [ ]
- [ ]
- [ ]
- [ ]

blackHatMonkey commented:
This is list of all the stuff that needs to be reviewed. @anakin1028 feel free to update it as we go.

- [ ] test_exception
- [ ] constants
- [ ] extensions
- [ ] deauth
- [ ] lure10
- [ ] pywifiphisher

blackHatMonkey commented on tests/test_extensions.py:
```diff
@@ -0,0 +1,132 @@
+import mock
+import shutil
+
+dir_of_executable = os.path.dirname(__file__)
```
We no longer need these since `tox` will take care of it.

sophron replied:
We want a user to be able to run a test independently w/o tox.

blackHatMonkey replied:
Not sure why would a user want to run the tests. The only people who should run the test are the developers. The developers should run tox on each change to make sure nothing is broken prior to making a PR.

sophron replied:
A developer (i.e. me) may want to run a test solely to verify it is passing. 

It's only a line. :)

blackHatMonkey replied:
Even if you didn't like `tox` you would run the tests with `nosetest` or `pytest` because you don't want to run a single test file. You would want to test all the test files to make sure nothing is broken. 

>It's only a line. :)

Technically 4 lines 😉 . But it's about consistency through out the project.

sophron replied:
> Even if you didn't like tox you would run the tests with nosetest or pytest because you don't want to run a single test file.

If I implement a feature on `ExtensionManager`, I don't want to run other tests. Currently, all of our tests are pretty quick, but in the future, they make take a number of seconds to finish.

blackHatMonkey replied:
You can still use `py.test tests/test_interfaces.py` to only run a specific file.

anakin1028 commented on wifiphisher/extensions/lure10.py:
why the channel is `["*"]` shouldn't that be `[7]` ?

anakin1028 commented on wifiphisher/extensions/lure10.py:
I cannot understand when I first see the `first` and `first_output` variable.  Maybe `has_beacon_forged` and `has_beacon_output` will be better? What do you think?

anakin1028 commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,305 @@
+                    try:
+                        self._socket.send(pkt)
+                    except BaseException:
+                        continue
```
and for the following `time.sleep(1)`. since it's may not be straightforward why we want to do a sleep here.

blackHatMonkey commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,305 @@
+                # Append only new channels (no duplicates)
+                self._channels_to_hop += list(set(channels_interested) -
+                                              set(self._channels_to_hop))
+
```
Although it is more compact compared to the extended version I think it's harder to read at the first glance but that's just me.

sophron replied:
It doesn't make a difference to me. If time-wise it makes a difference, we should change it.

sophron commented on wifiphisher/extensions/lure10.py:
The asterisk (*) means that this module will run in all channels registered by other modules. 

For example, if the module A registers channels 2 and 3 for sending data and module B registers channel 5, the lure10 module will send data to channels 2, 3 and 5. 

Yes, I need to document this somewhere.

sophron commented:
Guys, please do a final check, I will merge this during the weekend.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
`pkt` is never used here. Perhaps it should be removed.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
`bssid` is never used and it could be replaced by `_` since we don't care about it.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
@sophron Can you add some comments so it's more clear what's going on.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
I believe these two if conditions can be merged so it's a bit simpler.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
the `\` are redundant and should be removed.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
perhaps we can use `format` to make it easier to read.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
Forgot to add this is unreachable code.

anakin1028 commented on wifiphisher/common/extensions.py:
```diff
@@ -0,0 +1,294 @@
+        if self._channelhop_thread.is_alive():
+            self._send_thread.join(3)
+        self._socket.close()
+
```
`self._socket.close` will cause problem when `-nJ` option is used. Since it is not being initialized when there is no jamming.

anakin1028 commented on wifiphisher/extensions/lure10.py:
got it : )

anakin1028 commented on wifiphisher/pywifiphisher.py:
`deauthentication` can be removed now since we have `extension manager` right ?

sophron commented on wifiphisher/extensions/lure10.py:
It can't be removed. The API provides specific data, it's up to the module what to use or not.

sophron commented on wifiphisher/extensions/lure10.py:
It is not. `ExtensionManager` keeps running this method. The first time only it will output a message, the rest it will do nothing.

sophron commented on wifiphisher/extensions/lure10.py:
Let's be strict with PEP8 as far as coding style is concerned. I'm afraid I can't guarantee (time-wise) to be consistent with a coding style further than PEP8. But feel free to commit and fix them yourself if you want.

sophron commented:
FYI, pushed a couple more commits.

blackHatMonkey commented on wifiphisher/extensions/lure10.py:
But this statement is after the return. It will never be run.

blackHatMonkey commented:
@sophron Would it be ok if I moved the branch from your fork to the main repository so it is easier to modify and review?

sophron commented:
@blackHatMonkey Sure, go ahead!

Let's try to merge this within the next few days. We'll be able to implement many cool features afterwards.

blackHatMonkey commented:
Moved to #656.
